ddev/ddev
# Implement Django4 and python project types and nginx-gunicorn webserver_type, for #3709

## The Issue

* #3709 

Django4/gunicorn support

## How This PR Solves The Issue

* Add `django4` project type
* Add `nginx-gunicorn` webserver_type
* Autodetect project of type `django4` and set webserver_type in that situation
* Autodetect docroot == app for django4

## Manual Testing Instructions

Download the appropriate binaries from this PR and put them in your $PATH. Use `ddev --version` to verify that you have an appropriate version. See [Testing a PR](https://ddev.readthedocs.io/en/latest/developers/building-contributing/#testing-a-pr) for more detail.

- [ ] Check out a django4 app
- [ ] `ddev config` or `ddev config --auto` - the project type and webserver_type should be set
- [ ] You may need DJANGO_SETTINGS_MODULE environment variable set in case DDEV can't figure out where the settings are, as in the case of bakerydemo
- [ ] `ddev start` and do whatever is needed to import a db, etc.
- [ ] See if your project works

Some candidates for testing:
- [x] https://github.com/wagtail/bakerydemo, requires `DJANGO_SETTINGS_MODULE=bakerydemo.settings.local`
- [x] https://github.com/django-cms/django-cms-quickstart 
- [x] https://github.com/wsvincent/djangox
- [x] https://github.com/platformsh-templates/django4
- [x] https://github.com/greyli/sayhello
- [x] https://github.com/app-generator/flask-volt-dashboard


## TODO

- [x] Allow WSGI_APP variable to be set to override settings search and just launch that instead of getting it from settings.
- [x] current gunicorn WSGI_APP is problematic. Currently we're just using the name of the directory where we find manage.py, but how can we do it better?
- [ ] Experiment with not showing PHP-related things in config.yaml (php_version, composer_version, xdebug_enabled) if we have a python-type app. This may have to be deferred.
- [x] Add a project type of `python`, generic python (like `php` is generic php)
- [x] venv should be just named that and setup should be included in /etc/bashrc.d. venv should probably be gitignored, consider putting it into .ddev
- [x] Add settings.ddev.py, consider putting it in .ddev
- [ ] Is `docroot` naming (currently the app directory) going to be a problem for python folks?
- [x] Add `ddev python` command
- [x] Override database type during config to default to postgres:14 ?
- [x] venv support via pip. 
  - [x] Add to homeadditions/.bashrc.d/
  - [x] Do setup at start.sh (.web-entrypoint.d) time I guess
  - [x] pip install -r requirements.txt
  - [x] Initial setup should pip install gunicorn and add psycopg2
- [x] Package support: Need to install requirements.txt or pyproject.toml
- [x] healthcheck update
- [x] build-essential adds a *lot* to ddev-webserver. Can we live without it?
- [x] Make mailhog work
    - [x] Basic usage
    - [x] How to add extra gunicorn configuration
    - [x] `ddev python` command
    - [x] What happens when no app is discovered
    - [x] Places that talk just about PHP (like README.md) need to be updated
    - [x] All the places that talk about project type and webserver_type
    - [x] Mutagen not supported (yet?)
    - [x] Requirements.txt/toml behavior (ddev logs -f), cached pip
    - [x] psycopg2 may need build-essential (but psycopg2-binary is automatically installed and does not)

## Automated Testing Overview

Needs:
- [x] TestDdevFullSiteSetup stanza
- [x] TestWebserverType support for python/django
- [x] TestAppDetection support
- [x] `ddev python` command support test (TestCustomCommands?)
- [x] Testing support for pre-command
- [x] Testing support for WebEnvironment



<a href="https://gitpod.io/#https://github.com/ddev/ddev/pull/4768"><img src="https://gitpod.io/button/open-in-gitpod.svg"/></a>
